src/mpid/ch4/netmod/ofi/ofi_am_lmt.c

```diff
@@ -23,76 +23,76 @@ void MPIDI_OFI_am_lmt_init(void)
     MPIDIG_am_reg_cb(MPIDI_OFI_AM_LMT_ACK, NULL, &ofi_handle_lmt_ack);
 }
 
+/* declare static functions */
+static int ofi_do_rdma_read(void *dst, uint64_t src, size_t data_sz,
+                            MPIR_Context_id_t context_id, int src_rank, MPIR_Request * rreq);
+
+static int ofi_do_lmt_ack(int rank, int context_id, MPIR_Request * sreq_ptr);
+
 /* MPIDI_OFI_AM_LMT_REQ */
 
+struct am_lmt_hdr {
+    /* payload rdma info */
+    MPIR_Context_id_t context_id;
+    int src_rank;
+    uint64_t src_offset;
+    uint64_t rma_key;
+    /* orig am info */
+    int handler_id;
+    MPI_Aint am_hdr_sz;
+    MPI_Aint data_sz;
+    MPIR_Request *sreq_ptr;
+};
+
 int MPIDI_OFI_am_lmt_send(int rank, MPIR_Comm * comm, int handler_id,
                           const void *am_hdr, size_t am_hdr_sz,
                           const void *data, MPI_Aint data_sz, MPIR_Request * sreq)
 {
     int mpi_errno = MPI_SUCCESS;
 
-    MPIDI_OFI_am_header_t *msg_hdr;
```

Comment:
A major intention is to show that this is just a typical ch4-layer active message protocol. The new struct and namespace change is to adopt a common ch4-am-protocol template. The work to unify of protocol template, unfortunately, is still pending in my open/draft PRs that needs to be pushed one piece at a time :smile: 

Here is the template:
A set of am protocol lives in a self-contained `.c` file or/and a `.h` file for expose entrance functions or inline them. A protocol is made of a set of chained-packets or "am" messages, each consists of a `send`, send-callback, target-callback, and target completion callback for asynchronous payload transfer. And each packet or message defines its `am_xxx_hdr`. Since this `hdr` is coordination between sender and target-callback, it should be kept internal. 

So this `protocol.c` will be laid out as sections, each section define one packet leg, and they all look uniform: first a struct `am_xxx_hdr` definition, then a sender function, then a sender-side completion callback, then its target-side callback and completion callback, plus accessory helper routines.

The sender function needs to assemble the am_xxx_hdr and send it off. I would like it to look as uniform as it can, i.e. use `am_hdr`, `p_hdr` to downplay its uniqueness.

The target callback or handler need to cast the header into the correct type, then process it, and call the sender of the next leg. Again I would like to make it as uniform as it can.

With the uniformity, it should reduce the cognitive load in reading and maintain these am protocols and reduce potential overlooks and mistakes. It'll be nice if we can quickly scan and zoom into the unique part of each protocol.

Ideally, the only thing needs to be exposed outside an am protocol is the sender function of the first leg, then its target callbacks will initiate additional legs. In another word, the entire protocol is self-contained. In practice, we may need to expose sender functions for internal legs if we use asynchronous operations. More importantly, each am protocol is independent of each other with its own header and protocol definitions. They can be moved to different layers or selected independently.

Reply:
So that's fine, but it would make more sense if the code had moved out to the CH4-layer.